core/src/main/java/com/yahoo/oak/EntryHashSet.java

```diff
@@ -0,0 +1,561 @@
+    static final int INVALID_FULL_HASH = 0; // because memory is initially zeroed
+    static final int DEFAULT_COLLISION_CHAIN_LENGTH = 3;
+    // HASH - the full hash index of this entry (one integer, all bits).
+    private static final int HASH_FIELD_OFFSET = 2;
+    private static final int FULL_HASH_BITS = 32; // Hash needs to be an integer
+    // One additional to the key and value references field, which keeps the full hash index of the entry
+    private static final int ADDITIONAL_FIELDS = 1;  // # of primitive fields in each item of entries array
+
+    // number of entries candidates to try in case of collision
+    private AtomicInteger collisionChainLength = new AtomicInteger(DEFAULT_COLLISION_CHAIN_LENGTH);
+
+    private final OakComparator<K> comparator;
+
+    // use reference codec to encode Full Hash Number integer (first) with its updates counter (second)
+    private final ReferenceCodec hashCodec = new ReferenceCodec(
+        FULL_HASH_BITS, // bits# to represent full hash number as integer, bits# to represent update
+        ReferenceCodec.INVALID_BIT_SIZE); // counter are calculated upon previous parameters (also int)
+
+    /*----------------- Constructor -------------------*/
+
+    /**
+     * Create a new EntryHashSet
+     * @param vMM   for values off-heap allocations and releases
+     * @param kMM off-heap allocations and releases for keys
+     * @param entriesCapacity how many entries should this EntryOrderedSet keep at maximum
+     * @param keySerializer   used to serialize the key when written to off-heap
+     */
+    EntryHashSet(MemoryManager vMM, MemoryManager kMM, int entriesCapacity, OakSerializer<K> keySerializer,
+        OakSerializer<V> valueSerializer, OakComparator<K> comparator) {
+        // We add additional field for the head (dummy) node
+        super(vMM, kMM, ADDITIONAL_FIELDS, entriesCapacity, keySerializer, valueSerializer);
+        this.comparator = comparator;
+    }
+
+
+
+    /********************************************************************************************/
+    /*---------- Methods for managing full hash entry indexes (package visibility) -------------*/
+
+    /**
+     * getFullHashNumber returns the full hash entry index (the number only!) of the entry
+     * given by entry index "ei". The method serves internal and external EntryHashSet users.
+     */
+    private int getFullHashNumber(int ei) {
+        assert isIndexInBound(ei);
+        long fullHashNumberField = getEntryFieldLong(ei, HASH_FIELD_OFFSET);
+        // extract the hash number (first) from the updates counter (second)
+        return hashCodec.getFirst(fullHashNumberField);
+    }
+
+    /**
+     * getFullHashEntryField returns the full hash entry index (the entire field including counter!)
+     * of the entry given by entry index "ei". The method serves internal and external EntryHashSet users.
+     */
+    private long getFullHashEntryField(int ei) {
```

Comment:
So is this going to be `geVersionedHash()`?

Reply:
This is related to the naming issue from the above. Skipping.